File: aoc-py/solutions/day14.py
from __future__ import annotations
"""
Day 14: Parabolic Reflector Dish

https://adventofcode.com/2023/day/14
"""
__all__ = ('Day14',)

from typing import ClassVar, TypeAlias, TYPE_CHECKING

from ..solution import Solution

if TYPE_CHECKING:
    Grid: TypeAlias = list[list[str]]

class Day14(Solution):
    NAME: ClassVar[str] = 'Parabolic Reflector Dish'

    def _transpose(self, grid: Grid) -> Grid:
        """Transposes the grid"""
        return [list(row) for row in zip(*grid)]

    def _reverse_rows(self, grid: Grid) -> Grid:
        """Reverses each row in the grid"""
        for row in grid:
            row.reverse()
        return grid

    def _tilt_lever(self, grid: Grid) -> Grid:
        """Pushes all rounded rocks to the left side (west) of the matrix"""
        for row in grid:
            for i, element in enumerate(row):
                # goes from left to right for each row and finds all rounded rocks
                if element == 'O':
                    # for each rounded rock, traverse backwards in the row
                    # and if the tile before it is empty we can swap and move the rock to the left
                    # we keep doing so until we hit a non empty tile
                    for t in range(i, 0, -1):
                        if row[t - 1] == '.':
                            # is empty, swap
                            row[t - 1], row[t] = row[t], row[t - 1]
                        else:
                            # non empty tile, we are done pushing it
                            break
        return grid

    def _tilt_north(self, grid: Grid) -> Grid:
        """Tilts the rocks north (up) (transpose first)"""
        grid = self._transpose(grid)
        grid = self._tilt_lever(grid)
        return self._transpose(grid)

    def _tilt_south(self, grid: Grid) -> Grid:
        """Tilts the rocks south (down)
        (same as north; but reverse the columns
        (which are rows after transposal))
        """
        grid = self._transpose(grid)
        grid = self._reverse_rows(grid)
        grid = self._tilt_lever(grid)
        grid = self._reverse_rows(grid)
        return self._transpose(grid)

    def _tilt_east(self, grid: Grid) -> Grid:
        """Tilts the rocks east (right)
        Same as the default (left) but reverses the rows
        """
        grid = self._reverse_rows(grid)
        grid = self._tilt_lever(grid)
        return self._reverse_rows(grid)

    def _cycle(self, grid: Grid) -> Grid:
        grid = self._tilt_north(grid)
        # west = left (no need for pre manipulation)
        grid = self._tilt_lever(grid)
        grid = self._tilt_south(grid)
        return self._tilt_east(grid)

    def _get_load(self, grid: Grid) -> int:
        """Calculates the load
        by counting the rounded boulders per row
        """
        n_rows = len(grid)
        return sum(
            row.count('O') * (n_rows - i)
            for i, row in enumerate(grid)
        )

    def part_one(self, inp: str) -> int:
        grid = [list(row) for row in inp.splitlines()]
        grid = self._tilt_north(grid)
        return self._get_load(grid)

    def part_two(self, inp: str) -> int:
        grid = [list(row) for row in inp.splitlines()]

        # `cycles` will contain the extra non-repeating terms
        # + 1 period
        cycles = [grid]
        # contains the index of the element that first begins the repeating sequence
        # also equal to the length of the extra non-repeating prefixing chunk
        start = 0
        while True:
            # generates next term
            next_term = self._cycle(cycles[-1])
            # this marks the end of our first cycle
            # since we have already seen `next_term` before when it was the start of the cycle
            if next_term in cycles:
                # set `start` to the index of the start of the first cycle
                start = cycles.index(next_term)
                break
            else:
                cycles.append(next_term)

        for _ in range(
            # the total length of the repeating chunk
            (1_000_000_000 - start)
            # modulo by the period of the sequence
            # (gets the remainder at the end that does not complete another full cycle)
            % (len(cycles) - start)
            # add back the non-repeating chunk's length
            + start
        ):
            grid = self._cycle(grid)
        return self._get_load(grid)

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 111339
        assert p2 == 93736

File: src/bin/day14.rs
//! Day 14: Parabolic Reflector Dish
//!
//! <https://adventofcode.com/2023/day/14>
use std::{collections::VecDeque, fmt::Display};
use aoc_2023::Solution;

pub struct Day14;

type Grid = Vec<Vec<u8>>;

impl Day14 {
    fn transpose(grid: &Grid) -> Grid {
        (0..grid[0].len())
            .map(|i| (0..grid.len())
                .map(|j| grid[j][i])
                .collect::<Vec<u8>>()
            )
            .collect::<Grid>()
    }

    fn reverse_rows(grid: &mut Grid) {
        for row in grid {
            row.reverse();
        }
    }

    fn tilt_lever(grid: &mut Grid) {
        for row in grid {
            for i in 0..row.len() {
                if row[i] != b'O' {
                    continue;
                }
                for t in (1..=i).rev() {
                    if row[t - 1] != b'.' {
                        break;
                    }
                    row.swap(t, t - 1);
                }
            }
        }
    }

    fn tilt_north(grid: &Grid) -> Grid {
        let mut grid = Self::transpose(grid);
        Self::tilt_lever(&mut grid);
        Self::transpose(&grid)
    }

    fn tilt_south(grid: &Grid) -> Grid {
        let mut grid = Self::transpose(grid);
        Self::reverse_rows(&mut grid);
        Self::tilt_lever(&mut grid);
        Self::reverse_rows(&mut grid);
        Self::transpose(&grid)
    }

    fn tilt_east(grid: &mut Grid) {
        Self::reverse_rows(grid);
        Self::tilt_lever(grid);
        Self::reverse_rows(grid);
    }

    fn cycle(grid: &Grid) -> Grid {
        let mut grid = Self::tilt_north(grid);
        Self::tilt_lever(&mut grid);
        grid = Self::tilt_south(&grid);
        Self::tilt_east(&mut grid);
        grid
    }

    fn get_load(grid: &Grid) -> usize {
        let n_rows = grid.len();

        #[allow(clippy::naive_bytecount)]
        grid
            .iter()
            .enumerate()
            .map(|(i, row)|
                row
                    .iter()
                    .filter(|&&tile| tile == b'O')
                    .count()
                * (n_rows - i)
            )
            .sum()
    }

    fn get_grid<T: Display>(inp: T) -> Grid {
        inp
            .to_string()
            .lines()
            .map(|line| line
                .as_bytes()
                .to_vec()
            )
            .collect::<Grid>()
    }

    pub fn part_one<T: Display>(&self, inp: T) -> usize {
        Self::get_load(
            &Self::tilt_north(
                &Self::get_grid(inp)
            )
        )
    }

    /// # Panics
    ///
    /// If the cycles vec is empty
    pub fn part_two<T: Display>(&self, inp: T) -> usize {
        let grid = Self::get_grid(inp);

        let mut cycles = VecDeque::from([grid]);
        let start = loop {
            let next_term = Self::cycle(cycles
                .back()
                .unwrap()
            );

            if let Some(index) = cycles
                .iter()
                .position(|term| term == &next_term)
            {
                break index;
            }
            cycles.push_back(next_term);
        };

        Self::get_load(
            &cycles[
                (1_000_000_000 - start)
                % (cycles.len() - start)
                + start
            ]
        )
    }
}

impl Solution for Day14 {
    const NAME: &'static str = "Parabolic Reflector Dish";

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 111_339);
        assert_eq!(p2, 93736);
    }
}

fn main() {
    aoc_2023::run_day(14, &Day14);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}